File: fractaltorch/umht.py
import math
import torch
import torch.nn as nn
from .utils import set_seed

class UnifiedMultiHeadTransformerLSTMCell(nn.Module):
    def __init__(self, input_size, hidden_size, heads=4, lstm_layers=1, mem_len=32, seed: int = 1234):
        super().__init__()
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.heads = heads
        self.lstm_layers = lstm_layers
        self.mem_len = mem_len

        # set seed for deterministic initialization (attempt to mirror C#/TorchSharp defaults)
        set_seed(seed)

        # single-step LSTMCell to match C# LSTMCell behavior
        self.lstm_cell = nn.LSTMCell(input_size, hidden_size)
        self.self_attn = nn.MultiheadAttention(hidden_size, heads)
        self.cross_attn = nn.MultiheadAttention(hidden_size, heads)

        # optional input projection not implemented here; caller should match sizes
        self.residual_project = nn.Linear(hidden_size, hidden_size)
        self.router_mlp = nn.Linear(hidden_size, 2)

        # CT-gate params
        self.bottleneck_dim = max(1, hidden_size // 4)
        self.W_ct_gate = nn.Linear(input_size + hidden_size, hidden_size)
        self.W_ct_compress = nn.Linear(hidden_size, self.bottleneck_dim)
        self.W_ct_expand = nn.Linear(self.bottleneck_dim, hidden_size)

        # use explicit dropout module to mirror C# functional.dropout behavior with same p
        self._dropout = nn.Dropout(p=0.2)

        # internal rolling memory (S,B,H)
        self.register_buffer("memK", None)
        self.register_buffer("memV", None)

        # initialize weights similarly to TorchSharp defaults (kaiming_uniform for linear weights, zeros for bias)
        self._init_weights()

    def _init_weights(self):
        # Kaiming uniform for linear layers
        for m in self.modules():
            if isinstance(m, nn.Linear):
                nn.init.kaiming_uniform_(m.weight, a=math.sqrt(5))
                if m.bias is not None:
                    fan_in, _ = nn.init._calculate_fan_in_and_fan_out(m.weight)
                    bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
                    nn.init.uniform_(m.bias, -bound, bound)
            elif isinstance(m, nn.LSTMCell):
                # default PyTorch LSTMCell init is already fine; for parity we can use xavier for weights
                for name, param in m.named_parameters():
                    if 'weight' in name:
                        nn.init.xavier_uniform_(param)
                    elif 'bias' in name:
                        nn.init.zeros_(param)

    def _ensure_batch(self, x):
        t = x
        if t.dim() == 1:
            t = t.unsqueeze(0)
        return t

    def forward_step(self, x, h, c, externalK=None, externalV=None):
        # x: [B, input_size] or [input_size]
        xIn = self._ensure_batch(x)  # [B, input_size]

        # Ensure h,c have batch dim
        h_in = self._ensure_batch(h).squeeze(0) if h.dim() == 2 and h.shape[0] == 1 else h
        c_in = self._ensure_batch(c).squeeze(0) if c.dim() == 2 and c.shape[0] == 1 else c

        # LSTMCell update (single step)
        h_lstm, c_next = self.lstm_cell(xIn, (h_in, c_in))  # both [B, H]

        # Write into rolling memory memK/memV as [S,B,H]
        cur = h_lstm.unsqueeze(0)  # [1,B,H]
        if getattr(self, 'memK', None) is None or self.memK is None:
            # create memory tensors on same device
            self.memK = cur.detach()
            self.memV = cur.detach()
        else:
            newK = torch.cat([self.memK, cur.detach()], dim=0)
            newV = torch.cat([self.memV, cur.detach()], dim=0)
            if newK.shape[0] > self.mem_len:
                start = newK.shape[0] - self.mem_len
                # keep last mem_len entries
                self.memK = newK[start: , :, :]
                self.memV = newV[start: , :, :]
            else:
                self.memK = newK
                self.memV = newV

        # SELF attention over rolling memory
        # q: [1,B,H], key/value: [S,B,H]
        qSelf = h_lstm.unsqueeze(0)
        # nn.MultiheadAttention expects (L, N, E)
        self_out, _ = self.self_attn(qSelf, self.memK, self.memV)
        self_out = self_out.squeeze(0)  # [B,H]

        # EXTERNAL attention if provided
        if externalK is not None and externalV is not None:
            qCross = h_lstm.unsqueeze(0)
            ext_out, _ = self.cross_attn(qCross, externalK, externalV)
            external_out = ext_out.squeeze(0)
            has_external = True
        else:
            external_out = torch.zeros_like(h_lstm)
            has_external = False

        # ROUTER over self vs external
        router_logits = self.router_mlp(h_lstm)  # [B,2]
        router_weights = torch.softmax(router_logits, dim=1)  # [B,2]
        wSelf = router_weights[:, 0].unsqueeze(1)  # [B,1]
        wExt = router_weights[:, 1].unsqueeze(1)   # [B,1]

        routed_attn = wSelf * self_out + wExt * external_out  # [B,H]

        # residual projection and combined routed
        routed = h_lstm + routed_attn + self.residual_project(routed_attn)

        # CT-GATE temporal summary and bottleneck
        if getattr(self, 'memK', None) is None or self.memK is None or self.memK.shape[0] == 0:
            temporal_summary = torch.zeros_like(h_lstm)
        else:
            temporal_summary = self.memK.mean(dim=0)  # [B,H]

        compressed_small = self.W_ct_compress(temporal_summary)  # [B, bottleneck]
        expanded = self.W_ct_expand(compressed_small)            # [B, H]

        cs_dim = compressed_small.shape[1]
        reps = (self.hidden_size + cs_dim - 1) // cs_dim
        compressed_expanded = compressed_small.repeat(1, reps)[:, :self.hidden_size]

        gate_input = torch.cat([xIn, routed], dim=1)  # [B, input+H]
        g = torch.sigmoid(self.W_ct_gate(gate_input))  # [B,H]

        hCT = torch.lerp(compressed_expanded, expanded, g)

        # Ensemble and final output
        stacked = torch.stack([h_lstm, routed, hCT], dim=0)  # [3, B, H]
        hFinal = stacked.mean(dim=0)  # [B,H]

        hFinal = self._dropout(hFinal)

        # Return exactly like C# forward_step: (output, h, c)
        # output: hFinal [B,H]
        # h: h_lstm [B,H]
        # c: c_next [B,H]
        return hFinal, h_lstm, c_next

    def ResetMemory(self):
        self.memK = None
        self.memV = None

File: deepseekx/UMHT.cs
using System;
using System.Collections.Generic;
using System.Threading;
using TorchSharp;
using TorchSharp;
using TorchSharp.Modules;
using TorchSharp.Modules;
using static System.Net.Mime.MediaTypeNames;
using static TorchSharp.torch;
using static TorchSharp.torch;

public class UnifiedMultiHeadTransformerLSTMCell : nn.Module
{
    // rolling self-attention memory
    private Tensor? memK;   // [S,B,H]
    private Tensor? memV;   // [S,B,H]
    private readonly int memLen = 32;

    // sizes
    public readonly int InputSize;
    public readonly int HiddenSize;
    public readonly int hiddenSize; // for compatibility
    public readonly int NumHeads;

    // core modules
    private readonly LSTMCell lstm;
    private readonly MultiheadAttention selfAttn;
    private readonly MultiheadAttention crossAttn;

    private readonly Linear? inputProject;   // x -> hiddenSize if needed
    private readonly Linear residualProject; // extra projection on routed attn
    private readonly Linear routerMLP;       // router over attention banks (self vs external)

    // CT-gate parameters
    private readonly Linear W_ct_gate;       // gate over [x_t, attn]
    private readonly Linear W_ct_compress;   // compress temporal summary to bottleneck
    private readonly Linear W_ct_expand;     // expand back to hiddenSize
    private readonly int bottleneckDim;

    // final output head
    public readonly nn.Module<Tensor, Tensor> output;

    public UnifiedMultiHeadTransformerLSTMCell(
        int inputSize,
        int hiddenSize,
        int numHeads,
        int outputSize = 1)
        : base("unified_multihead_transformer_lstm_cell_router_ct")
    {
        InputSize = inputSize;
        HiddenSize = hiddenSize;
        this.hiddenSize = hiddenSize;
        NumHeads = numHeads;

        // LSTM operates in hidden space after the first projection
        // but we keep inputSize for x_t; we project x_t to hiddenSize if needed
        lstm = nn.LSTMCell(inputSize, hiddenSize);

        // self-attention over rolling memory of hidden states
        selfAttn = nn.MultiheadAttention(hiddenSize, numHeads);
        // optional cross-attention over external memory (same hidden size)
        crossAttn = nn.MultiheadAttention(hiddenSize, numHeads);

        if (inputSize != hiddenSize)
            inputProject = nn.Linear(inputSize, hiddenSize);
        else
            inputProject = null;

        residualProject = nn.Linear(hiddenSize, hiddenSize);

        // router over two banks: self-mem and external-mem
        // routerMLP: [B,H] -> [B,2]
        routerMLP = nn.Linear(hiddenSize, 2);

        // CT-gate
        bottleneckDim = Math.Max(1, hiddenSize / 4);
        // gate sees original input x_t and the attention/routed representation
        W_ct_gate = nn.Linear(inputSize + hiddenSize, hiddenSize);
        W_ct_compress = nn.Linear(hiddenSize, bottleneckDim);
        W_ct_expand = nn.Linear(bottleneckDim, hiddenSize);

        this.output = nn.Sequential(
            nn.Dropout(0.1), // 10% Dropout für die finale Entscheidung
            nn.Linear(hiddenSize, outputSize) // Hier 'outputSize' statt 'vocabSize'
        );
        // Initialisierung für stabilen Gradientenfluss
        // Sicherer Zugriff auf die internen Parameter von MultiheadAttention
        foreach (var (name, param) in crossAttn.named_parameters())
        {
            if (name.Contains("in_proj_weight"))
            {
                nn.init.xavier_uniform_(param);
            }
            if (name.Contains("in_proj_bias"))
            {
                nn.init.zeros_(param);
            }
        }
        RegisterComponents();
    }

    // Backwards-compatible single-step forward without external banks.
    public (Tensor output, Tensor h, Tensor c) forward_step(Tensor x, Tensor h, Tensor c)
    {
        return forward_step(x, h, c, externalK: null, externalV: null);
    }

    // Router- and CT-gate-aware forward step.
    // x: [B,inputSize] or [inputSize] or [1,inputSize]
    // h,c: [B,HiddenSize]
    // externalK/externalV: [S_ext,B,HiddenSize] or null
    public (Tensor output, Tensor h, Tensor c) forward_step(
        Tensor x,
        Tensor h,
        Tensor c,
        Tensor? externalK,
        Tensor? externalV)
    {

        // Ensure batch dimension for x_t
        Tensor xIn = x;
        if (xIn.dim() == 1)
            xIn = xIn.unsqueeze(0); // [1,inputSize]

        // 1) LSTM update in input space -> h_lstm
        var (hLstm, cNext) = lstm.forward(xIn, (h, c)); // hLstm: [B,H]

        // Write hLstm into rolling self-attention memory
        var cur = hLstm.unsqueeze(0); // [1,B,H]

        if (memK is null || memV is null)
        {
            memK = cur.detach();
            memV = cur.detach();
        }
        else
        {
            var newK = torch.cat(new[] { memK, cur.detach() }, 0); // [S_old+1,B,H]
            var newV = torch.cat(new[] { memV, cur.detach() }, 0);

            if (newK.shape[0] > memLen)
            {
                var start = newK.shape[0] - memLen;
                memK.Dispose();
                memV.Dispose();
                memK = newK.slice(0, start, newK.shape[0], 1);
                memV = newV.slice(0, start, newV.shape[0], 1);
            }
            else
            {
                memK.Dispose();
                memV.Dispose();
                memK = newK;
                memV = newV;
            }
        }

        // 2) SELF-ATTENTION over rolling memory
        Tensor selfOut;
        {
            // memK/memV: [S,B,H]
            var qSelf = hLstm.unsqueeze(0); // [1,B,H]
            var selfRes = selfAttn.forward(qSelf, memK!, memV!, null, false, null);
            selfOut = selfRes.Item1.squeeze(0); // [B,H]
        }

        // 3) OPTIONAL EXTERNAL ATTENTION
        Tensor externalOut;
        bool hasExternal = (externalK is not null) && (externalV is not null);

        if (hasExternal)
        {
            var qCross = hLstm.unsqueeze(0); // [1,B,H]
            var crossRes = crossAttn.forward(qCross, externalK!, externalV!, null, false, null);
            externalOut = crossRes.Item1.squeeze(0); // [B,H]
        }
        else
        {
            // Use crossAttn over the internal memory as fallback so crossAttn parameters
            // are part of the graph and receive gradients (prevents zero-grad diagnostics).
            // This mirrors self-attention but goes through the separate crossAttn module.
            var qCross = hLstm.unsqueeze(0); // [1,B,H]
            var crossRes = crossAttn.forward(qCross, memK!, memV!, null, false, null);
            externalOut = crossRes.Item1.squeeze(0); // [B,H]
        }

        // 4) ROUTER over self vs external
        var routerLogits = routerMLP.forward(hLstm);   // [B,2]
        var routerWeights = routerLogits.softmax(1);   // [B,2]

        var wSelf = routerWeights.index(new TensorIndex[] {
            TensorIndex.Colon,
            TensorIndex.Single(0)
        }).unsqueeze(1); // [B,1]

        var wExt = routerWeights.index(new TensorIndex[] {
            TensorIndex.Colon,
            TensorIndex.Single(1)
        }).unsqueeze(1); // [B,1]

        // For debugging, keep this if you want to see routing behavior:
        // Console.WriteLine($"router: self={wSelf.mean().ToSingle():F3} ext={wExt.mean().ToSingle():F3}");

        var routedAttn = wSelf * selfOut + wExt * externalOut; // [B,H]

        // residual projection on routed attention
        var routed = hLstm + routedAttn + residualProject.forward(routedAttn); // [B,H]

        // 5) CT-GATE (temporal transform gate)

        // Temporal summary: mean over time dimension of memK [S,B,H] -> [B,H]
        Tensor temporalSummary;
        if (memK is null || memK.shape[0] == 0)
        {
            temporalSummary = torch.zeros_like(hLstm);
        }
        else
        {
            temporalSummary= memK.mean(new long[] { 0L });
        }

        // Compress summary to bottleneck and expand back
        var compressedSmall = W_ct_compress.forward(temporalSummary); // [B,bottleneck]
        var expanded = W_ct_expand.forward(compressedSmall);          // [B,H]

        // Expand compressedSmall to hidden size for blending
        int csDim = (int)compressedSmall.shape[1];
        int reps = (HiddenSize + csDim - 1) / csDim;

        var compressedSmallExpanded =
            compressedSmall
                .repeat(new long[] { 1L, (long)reps })   // repeat needs long[]
                .narrow(1L, 0L, (long)HiddenSize);       // narrow needs long args

        // Gate uses original x_t and routed attention representation
        var gateInput = torch.cat(new Tensor[] { xIn, routed }, 1); // [B, input+H]
        var g = torch.sigmoid(W_ct_gate.forward(gateInput));        // [B,H]

        var hCT = torch.lerp(compressedSmallExpanded, expanded, g); // [B,H]

        // 6) Final blend (ensemble of LSTM, route
        // [B,H]

        // In UnifiedMultiHeadTransformerLSTMCell.forward
        // Ersetzen Sie den finalen Blend-Abschnitt:

        var ensemble = new List<Tensor> { hLstm, routed, hCT };
        using var stacked = torch.stack(ensemble, 0);
        var hFinal = stacked.mean(new long[] { 0 });

        // NEU: LayerNorm vor dem Dropout stabilisiert die Skalierung für den Bias
        hFinal = torch.nn.functional.layer_norm(hFinal, new long[] { HiddenSize });

        hFinal = nn.functional.dropout(hFinal, 0.2, this.training);

        // Korrektur der Rückgabewerte:
        // hLstm ist der neue Hidden-State (h), cNext ist der neue Cell-State (c)
        return (hFinal, hLstm, cNext);

    }

    public void ResetMemory()
    {
        memK?.Dispose();
        memV?.Dispose();
        memK = null;
        memV = null;
    }
}
